File: Typescript/src/hasPathSum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

import { TreeNode } from "./TreeNode";

export function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
	if (!root) return false;
	if (!root.left && !root.right) {
		return root.val === targetSum;
	}
	const newTargetSum = targetSum - root.val;
	return hasPathSum(root.left, newTargetSum) || hasPathSum(root.right, newTargetSum);
};

File: C/exercise/hasPathSum.c

#include "hasPathSum.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool hasPathSum(struct TreeNode* root, int targetSum) {
	if (root == NULL) {
		return false;
	}
	targetSum -= root->val;
	if (root->left == NULL && root->right == NULL) {
		return targetSum == 0;
	}
	return hasPathSum(root->left, targetSum) || hasPathSum(root->right, targetSum);
}
